lib/input/sftp.go

```diff
@@ -0,0 +1,495 @@
+			r, err := NewSFTP(conf.SFTP, log, stats)
+			if err != nil {
+				return nil, err
+			}
+			return NewAsyncReader(
+				TypeSFTP,
+				true,
+				reader.NewAsyncBundleUnacks(
+					reader.NewAsyncPreserver(r),
+				),
+				log, stats,
+			)
+		},
+		Status: docs.StatusExperimental,
+		Summary: `
+Downloads objects via an SFTP connection.`,
+		Description: `
+Downloads objects via an SFTP connection.
+## Metadata
+This input adds the following metadata fields to each message:
+` + "```" + `
+- sftp_file_path
+` + "```" + `
+You can access these metadata fields using [function interpolation](/docs/configuration/interpolation#metadata).`,
+
+		FieldSpecs: docs.FieldSpecs{
+			docs.FieldCommon(
+				"server",
+				"The server to connect to that has the target files.",
+			),
+			docs.FieldCommon(
+				"port",
+				"The port to connect to on the server.",
+			),
+			docs.FieldCommon(
+				"credentials",
+				"The credentials to use to log into the server.",
+			).WithChildren(credentialsFields...),
+			docs.FieldCommon(
+				"filename",
+				"The name of the file to pull messages from. If not provided, all the files in the path will be processed.",
+			),
+			docs.FieldCommon(
+				"path",
+				"The path of the directory or file that it will process.",
+			),
+			docs.FieldCommon(
+				"max_connection_attempts",
+				"How many times it will try to connect to the server before exiting with an error.",
+			),
+			docs.FieldAdvanced(
+				"retry_sleep_duration",
+				"How long it will sleep after failing to connect to the server before trying again, defaults to 5s if not provided.",
+				"10s", "5m",
+			),
+			codec.ReaderDocs,
+			docs.FieldAdvanced("delete_objects", "Whether to delete files from the server once they are processed."),
+		},
+		Categories: []Category{
+			CategoryServices,
+			CategoryNetwork,
+		},
+	}
+}
+
+//------------------------------------------------------------------------------
+
+// SFTPConfig contains configuration fields for the SFTP input type.
+type SFTPConfig struct {
+	Server                string          `json:"server" yaml:"server"`
+	Port                  int             `json:"port" yaml:"port"`
+	Filename              string          `json:"filename" yaml:"filename"`
+	Credentials           SFTPCredentials `json:"credentials" yaml:"credentials"`
+	Path                  string          `json:"path" yaml:"path"`
+	MaxConnectionAttempts int             `json:"max_connection_attempts" yaml:"max_connection_attempts"`
+	RetrySleepDuration    string          `json:"retry_sleep_duration" yaml:"retry_sleep_duration"`
+	Codec                 string          `json:"codec" yaml:"codec"`
+	DeleteObjects         bool            `json:"delete_objects" yaml:"delete_objects"`
+}
+
+// SFTPCredentials contains the credentials for connecting to the SFTP server
+type SFTPCredentials struct {
+	Username string `json:"username" yaml:"username"`
+	Secret   string `json:"secret" yaml:"secret"`
```

Comment:
I might be wrong but I think it's probably more common for this to be referred to as "password" in the world of ssh/sftp.